Added support for processor affinity for afl-fuzz workers

yoava333 opened the pull request:
this patch adds support for affinity & fixes https://github.com/ivanfratric/winafl/issues/91

ivanfratric commented on afl-fuzz.c:
```diff
@@ -383,6 +385,151 @@ static void shuffle_ptrs(void** ptrs, u32 cnt) {
+
+static void bind_to_free_cpu(void) {
+
+  u8 cpu_used[4096];
```
cpu_used[64]?

ivanfratric commented on afl-fuzz.c:
```diff
@@ -383,6 +385,151 @@ static void shuffle_ptrs(void** ptrs, u32 cnt) {
 
 }
 
+static void usleep(u64 usec)
```
Is this really needed? Why won't the plain old Sleep() work, considering the intervals involved are not sufficiently small to justify needing microsecond precision.

yoava333 commented:
Addressed your comments for the code.

I ran two benchmarks:
* on a CPU E5-2630 v4 @ 2.20GHz (20 core, HT) - I saw ~10% increase - from ~450/s to ~500/s
* on a my laptop i7-7700HQ (8 cores, HT):
  - 2 fuzzers without affinity: ~1000/s
  - 2 fuzzers with affinity to Hyper threaded cores: ~950/s, ~650/s
  - 2 fuzzers with affinity to different hyper threaded cores: ~1180/s (~18% increase)
  - 2 fuzzers on the same core ~480/s

So it depends if you are using all your cores, if not this might degrade performance since sequential cores are usually hyper threaded cores.

ivanfratric commented:
That's an interesting observation about sequential cores. As a workaround, do you think it makes sense to first try to pin to even cores (0, 2, 4, ...) and if none are available do a pass over odd cores?

yoava333 commented:
I think it does make sense, it doesn't hurt and in my tests (on both machines) it showed better performance when using subset of my cores. I've updated the code.

ivanfratric commented:
Thanks!

0vercl0k commented:
This is super cool btw - never expected to have an 18% gap between both
configurations :). Thanks @ivanfratric and @yoava333!

Cheers

2018-03-19 5:00 GMT-07:00 ivanfratric <notifications@github.com>:

> Thanks!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ivanfratric/winafl/pull/92#issuecomment-374188524>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABaHRbQgM7l3UaCsHdnMkdkE4-AOycMdks5tf53MgaJpZM4SrNrR>
> .
>
